support-cron-scheduler crashed while updating job and the original REST action address is unreachable

<!--🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅

Hello there! 😄

To expedite issue processing please search open and closed issues before submitting a new one.
Existing issues often contain information about workarounds, resolution, or progress updates.

Please do not remove sections of this template when submitting your issue.  If the section is not relevant to your issue, leave the section but enter "n/a" for not applicable.

Note some sections are marked "REQUIRED" and must be provided.  Never mark required sections "n/a".

🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅-->


# 🐞 Bug Report

### Affected Services [**REQUIRED**]
<!-- Can you pin-point one or more EdgeX services (Data, Metadata, Command, etc...) as the source of the bug? -->
<!-- ✍️edit: --> The issue is located in: support-cron-scheduler


### Is this a regression?

<!-- Did this behavior work correctly in the previous version? -->
<!-- ✍️--> No

### Description and Minimal Reproduction [**REQUIRED**]
1. Create a job with REST action and the address is unreachable
```
POST /job

[
  {
    "apiVersion": "v3",
    "scheduleJob": {
      "name": "test_job_3",
      "definition": {
        "type": "INTERVAL",
        "startTimestamp": 1725502020000,
        "endTimestamp": 1725502200000,
        "interval": "20s"
      },
      "actions": [
        {
          "type": "REST",
          "contentType": "application/json",
          "address": "http://edgex-core-data:59880/api/v3/ping",
          "method": "GET"
        }
      ],
      "labels": ["core-data"]
    }
  }
]
```
2. Update the job when the job is started
```
PATCH /job

[
    {
        "apiVersion": "v3",
        "scheduleJob": {
            "name": "test_job_1",
            "actions": [
                {
                "type": "REST",
                "contentType": "application/json",
                "address": "http://172.17.0.1:59880/api/v3/ping",
                "method": "GET"
                }
            ]
        }
    }
]
```


## 🔥 Exception or Error
<pre><code>
level=INFO ts=2024-09-05T02:06:19.934494174Z app=support-cron-scheduler source=manager.go:60 msg="New scheduled job test_job_3 was added into the scheduler manager. ScheduleJob ID: , Correlation-ID: 7dd0726f-9835-413a-9e89-56ab244501e3"
level=DEBUG ts=2024-09-05T02:06:19.938491803Z app=support-cron-scheduler source=schedulejob.go:230 msg="The scheduled job: test_job_3 will be started at 2024-09-05 10:07:00 +0800 CST (timestamp: 1725502020000). Correlation-ID: 7dd0726f-9835-413a-9e89-56ab244501e3"
level=DEBUG ts=2024-09-05T02:06:19.938558387Z app=support-cron-scheduler source=schedulejob.go:246 msg="The scheduled job: test_job_3 will be stopped at 2024-09-05 10:10:00 +0800 CST (timestamp: 1725502200000). Correlation-ID: 7dd0726f-9835-413a-9e89-56ab244501e3"
level=DEBUG ts=2024-09-05T02:06:19.938577554Z app=support-cron-scheduler source=schedulejob.go:49 msg="Successfully created the scheduled job. ScheduleJob ID: 9940b66b-f62e-49d5-925c-d4fc0d6f9259, Correlation-ID: 7dd0726f-9835-413a-9e89-56ab244501e3"
level=DEBUG ts=2024-09-05T02:06:35.904304878Z app=support-cron-scheduler source=reporter.go:195 msg="Publish 0 metrics to the 'edgex/telemetry/support-cron-scheduler' base topic"
level=DEBUG ts=2024-09-05T02:06:35.904697383Z app=support-cron-scheduler source=manager.go:123 msg="Reported metrics..."
level=DEBUG ts=2024-09-05T02:07:00.000648766Z app=support-cron-scheduler source=manager.go:113 msg="The scheduled job test_job_3 was started. Correlation-ID: 7dd0726f-9835-413a-9e89-56ab244501e3"
level=DEBUG ts=2024-09-05T02:07:05.903573244Z app=support-cron-scheduler source=reporter.go:195 msg="Publish 0 metrics to the 'edgex/telemetry/support-cron-scheduler' base topic"
level=DEBUG ts=2024-09-05T02:07:05.903826249Z app=support-cron-scheduler source=manager.go:123 msg="Reported metrics..."
level=DEBUG ts=2024-09-05T02:07:19.803766626Z app=support-cron-scheduler source=manager.go:101 msg="The scheduled job test_job_3 was stopped and removed from the scheduler manager. Correlation-ID: 0c43a277-3937-4b9f-b153-422a7596078c"
level=DEBUG ts=2024-09-05T02:07:19.80388967Z app=support-cron-scheduler source=manager.go:75 msg="The old scheduled job test_job_3 was removed from the scheduler manager while updating it. ScheduleJob ID: 9940b66b-f62e-49d5-925c-d4fc0d6f9259, Correlation-ID: 0c43a277-3937-4b9f-b153-422a7596078c"
level=DEBUG ts=2024-09-05T02:07:19.804141382Z app=support-cron-scheduler source=manager.go:81 msg="Scheduled job test_job_3 was updated into the scheduler manager. ScheduleJob ID: 9940b66b-f62e-49d5-925c-d4fc0d6f9259, Correlation-ID: 0c43a277-3937-4b9f-b153-422a7596078c"
level=DEBUG ts=2024-09-05T02:07:19.808064075Z app=support-cron-scheduler source=schedulejob.go:227 msg="The startTimestamp is expired for the scheduled job: test_job_3, which will be started immediately. Correlation-ID: 0c43a277-3937-4b9f-b153-422a7596078c"
level=DEBUG ts=2024-09-05T02:07:19.808136451Z app=support-cron-scheduler source=schedulejob.go:246 msg="The scheduled job: test_job_3 will be stopped at 2024-09-05 10:10:00 +0800 CST (timestamp: 1725502200000). Correlation-ID: 0c43a277-3937-4b9f-b153-422a7596078c"
level=DEBUG ts=2024-09-05T02:07:19.808160868Z app=support-cron-scheduler source=schedulejob.go:133 msg="Successfully patched the scheduled job: test_job_3. ScheduleJob ID: 9940b66b-f62e-49d5-925c-d4fc0d6f9259, Correlation-ID: 0c43a277-3937-4b9f-b153-422a7596078c"
level=DEBUG ts=2024-09-05T02:07:19.808281079Z app=support-cron-scheduler source=manager.go:113 msg="The scheduled job test_job_3 was started. Correlation-ID: 0c43a277-3937-4b9f-b153-422a7596078c"
level=DEBUG ts=2024-09-05T02:07:35.902658273Z app=support-cron-scheduler source=reporter.go:195 msg="Publish 0 metrics to the 'edgex/telemetry/support-cron-scheduler' base topic"
level=DEBUG ts=2024-09-05T02:07:35.902993112Z app=support-cron-scheduler source=manager.go:123 msg="Reported metrics..."
level=DEBUG ts=2024-09-05T02:08:05.905744471Z app=support-cron-scheduler source=reporter.go:195 msg="Publish 0 metrics to the 'edgex/telemetry/support-cron-scheduler' base topic"
level=DEBUG ts=2024-09-05T02:08:05.905924433Z app=support-cron-scheduler source=manager.go:123 msg="Reported metrics..."
level=DEBUG ts=2024-09-05T02:08:09.81604684Z app=support-cron-scheduler source=gocron.go:102 msg="Failed to execute the rest action: fail to send the HTTP request -> Get \"http://172.17.0.8:59880/api/v3/ping\": dial tcp 172.17.0.8:59880: i/o timeout"
level=ERROR ts=2024-09-05T02:08:09.820227425Z app=support-cron-scheduler source=manager.go:255 msg="failed to add a new schedule action record for job: test_job_3, Correlation-ID: 2e7131e9-88b0-4908-9962-0894882e6797, err: failed to insert schedule action record: ERROR: invalid input syntax for type uuid: \"\" (SQLSTATE 22P02) "
level=DEBUG ts=2024-09-05T02:08:29.817082896Z app=support-cron-scheduler source=gocron.go:102 msg="Failed to execute the rest action: fail to send the HTTP request -> Get \"http://172.17.0.8:59880/api/v3/ping\": dial tcp 172.17.0.8:59880: i/o timeout"
level=ERROR ts=2024-09-05T02:08:29.820176129Z app=support-cron-scheduler source=manager.go:255 msg="failed to add a new schedule action record for job: test_job_3, Correlation-ID: 2e7131e9-88b0-4908-9962-0894882e6797, err: failed to insert schedule action record: ERROR: invalid input syntax for type uuid: \"\" (SQLSTATE 22P02) "
level=DEBUG ts=2024-09-05T02:08:35.903689514Z app=support-cron-scheduler source=reporter.go:195 msg="Publish 0 metrics to the 'edgex/telemetry/support-cron-scheduler' base topic"
level=DEBUG ts=2024-09-05T02:08:35.904050605Z app=support-cron-scheduler source=manager.go:123 msg="Reported metrics..."
level=DEBUG ts=2024-09-05T02:08:49.81446939Z app=support-cron-scheduler source=gocron.go:102 msg="Failed to execute the rest action: fail to send the HTTP request -> Get \"http://172.17.0.8:59880/api/v3/ping\": dial tcp 172.17.0.8:59880: i/o timeout"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x7e5548]

goroutine 547 [running]:
github.com/edgexfoundry/edgex-go/internal/support/cronscheduler/infrastructure.(*manager).addNewJob.func2({0xdc, 0xa5, 0xa2, 0x43, 0x2c, 0xde, 0x4f, 0x3d, 0xa6, 0x27, ...}, ...)
        github.com/edgexfoundry/edgex-go/internal/support/cronscheduler/infrastructure/manager.go:221 +0x108
reflect.Value.call({0xfa5ea0?, 0x4000207560?, 0xee3690430ca8?}, {0x12856b4, 0x4}, {0x40003ccb90, 0x3, 0x0?})
        reflect/value.go:596 +0x994
reflect.Value.Call({0xfa5ea0?, 0x4000207560?, 0x400015f378?}, {0x40003ccb90?, 0xee369040eac8?, 0xee36d70795b8?})
        reflect/value.go:380 +0x94
github.com/go-co-op/gocron/v2.callJobFuncWithParams({0xfa5ea0?, 0x4000207560?}, {0x400015f608, 0x3, 0x897c6f688daa27a6?})
        github.com/go-co-op/gocron/v2@v2.11.0/util.go:29 +0x1fc
github.com/go-co-op/gocron/v2.(*executor).runJob(_, {{0x14b4858, 0x40004ac3c0}, 0x400035a9c0, {0xdc, 0xa5, 0xa2, 0x43, 0x2c, 0xde, ...}, ...}, ...)
        github.com/go-co-op/gocron/v2@v2.11.0/executor.go:404 +0x63c
github.com/go-co-op/gocron/v2.(*executor).start.func1.1({{0x14b4858, 0x40004ac3c0}, 0x400035a9c0, {0xdc, 0xa5, 0xa2, 0x43, 0x2c, 0xde, 0x4f, ...}, ...})
        github.com/go-co-op/gocron/v2@v2.11.0/executor.go:221 +0x64
created by github.com/go-co-op/gocron/v2.(*executor).start.func1 in goroutine 546
        github.com/go-co-op/gocron/v2@v2.11.0/executor.go:220 +0x4fc

</code></pre>


## 🌍  Your Environment
**Deployment Environment:** Ubuntu 2204/arm64

**EdgeX Version [**REQUIRED**]:** Main


**Anything else relevant?**

--- a/internal/support/cronscheduler/application/schedulejob.go
+++ b/internal/support/cronscheduler/application/schedulejob.go
@@ -119,6 +119,11 @@ func PatchScheduleJob(ctx context.Context, dto dtos.UpdateScheduleJob, dic *di.C
 
 	requests.ReplaceScheduleJobModelFieldsWithDTO(&job, dto)
 
+	// Add the ID for each action, the old actions will be replaced by the new actions
+	for i, action := range job.Actions {
+		job.Actions[i] = action.WithId("")
+	}
+
 	err = schedulerManager.UpdateScheduleJob(job, correlationId)
 	if err != nil {
 		return errors.NewCommonEdgeXWrapper(err)

--- a/internal/support/cronscheduler/controller/http/schedulejob_test.go
+++ b/internal/support/cronscheduler/controller/http/schedulejob_test.go
@@ -101,7 +101,7 @@ func TestAddScheduleJob(t *testing.T) {
 
 	valid := addScheduleJobRequestData()
 	model := dtos.ToScheduleJobModel(valid.ScheduleJob)
-	dbClientMock.On("AddScheduleJob", context.Background(), mock.Anything).Return(model, nil)
+	dbClientMock.On("AddScheduleJob", context.Background(), mock.MatchedBy(func(job models.ScheduleJob) bool { return job.Name == testScheduleJobName })).Return(model, nil)
 	schedulerManagerMock.On("AddScheduleJob", mock.MatchedBy(func(job models.ScheduleJob) bool { return job.Name == testScheduleJobName }), testCorrelationID).Return(nil)
 	schedulerManagerMock.On("StartScheduleJobByName", model.Name, testCorrelationID).Return(nil)
 
@@ -410,8 +410,8 @@ func TestPatchScheduleJob(t *testing.T) {
 
 	valid := testReq
 	dbClientMock.On("ScheduleJobById", context.Background(), *valid.ScheduleJob.Id).Return(model, nil)
-	dbClientMock.On("UpdateScheduleJob", context.Background(), model).Return(nil)
-	schedulerManagerMock.On("UpdateScheduleJob", model, testCorrelationID).Return(nil)
+	dbClientMock.On("UpdateScheduleJob", context.Background(), mock.MatchedBy(func(job models.ScheduleJob) bool { return job.Name == testScheduleJobName })).Return(nil)
+	schedulerManagerMock.On("UpdateScheduleJob", mock.MatchedBy(func(job models.ScheduleJob) bool { return job.Name == testScheduleJobName }), testCorrelationID).Return(nil)
 	validWithNoReqID := testReq
 	validWithNoReqID.RequestId = ""
 	validWithNoId := testReq

